sway-lib-std/src/storage.sw

```diff
@@ -638,3 +639,91 @@ impl<V> StorageVec<V> {
         store(__get_storage_key(), 0);
     }
 }
+
+pub struct StorageBytes {}
+
+impl StorageBytes {
+    /// Takes a `Bytes` type and stores the underlying collection of tightly packed bytes.
+    ///
+    /// ### Arguments
+    ///
+    /// * `bytes` - The bytes which will be stored.
+    ///
+    /// ### Examples
+    ///
+    /// ```sway
+    /// storage {
+    ///     stored_bytes: StorageBytes = StorageBytes {}
+    /// }
+    ///
+    /// fn foo() {
+    ///     let mut bytes = Bytes::new();
+    ///     bytes.push(5_u8);
+    ///     bytes.push(7_u8);
+    ///     bytes.push(9_u8);
+    ///
+    ///     storage.stored_bytes.write_bytes(bytes);
+    /// }
+    /// ```
+    #[storage(write)]
+    pub fn write_bytes(self, mut bytes: Bytes) {
+        // Get the number of storage slots needed based on the size of bytes.
+        let number_of_slots = (bytes.len() + 31) >> 5;
+
+        // The bytes capacity needs to be greater than or a multiple of 32 bytes so we can 
+        // make the 'quad' storage instruction store without accessing unallocated heap memory.
+        if bytes.buf.cap < number_of_slots * 32 {
+            bytes.buf.ptr = realloc_bytes(bytes.buf.ptr, bytes.buf.cap, number_of_slots * 32);
+        }
+
+        // Store `number_of_slots * 32` bytes starting at storage slot `key`.
+        let key = sha256(__get_storage_key());
+        let _ = __state_store_quad(key, bytes.buf.ptr, number_of_slots);
+
+        // Store the length of the bytes
+        store(__get_storage_key(), bytes.len());
+    }
+
+    /// Constructs a `Bytes` type from a collection of tightly packed bytes in storage.
+    ///
+    /// ### Examples
+    ///
+    /// ```sway
+    /// storage {
+    ///     stored_bytes: StorageBytes = StorageBytes {}
+    /// }
+    ///
+    /// fn foo() {
+    ///     let mut bytes = Bytes::new();
+    ///     bytes.push(5_u8);
+    ///     bytes.push(7_u8);
+    ///     bytes.push(9_u8);
+    ///     storage.stored_bytes.write_bytes(bytes);
+    ///
+    ///     let retrieved_bytes = storage.stored_bytes.into_bytes(key);
+    ///     assert(bytes == retrieved_bytes);
+    /// }
+    /// ```
+    #[storage(read)]
+    pub fn into_bytes(self) -> Bytes {
```

Comment:
I wonder if this method should instead return an `Option<Bytes>`, and let the developer decide if they want to create a new empty `Bytes` on the heap if the stored value is `None`?

Reply:
Also, it would be nice to reuse our existing traits for conversions where possible, i.e: the `From` trait has `from` and `into` methods that could maybe be implemented here for converting between `Bytes` and `StorageBytes`?
I'm not sure how good a fit that would be here though.